addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java

```diff
@@ -0,0 +1,150 @@
+import org.eclipse.smarthome.config.discovery.ExtendedDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.airvisualnode.AirVisualNodeBindingConstants;
+import org.openhab.binding.airvisualnode.internal.config.AirVisualNodeConfig;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import jcifs.netbios.NbtAddress;
+import jcifs.smb.SmbFile;
+
+/**
+ * Autodiscovery for AirVisual Node by searching for a host advertised with the NetBIOS name 'AVISUAL-<SerialNumber>'.
+ *
+ * @author Victor Antonovich - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true)
+public class AirVisualNodeDiscoveryService extends AbstractDiscoveryService implements ExtendedDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(AirVisualNodeDiscoveryService.class);
+
+    public static final String AVISUAL_WORKGROUP_NAME = "MSHOME";
+
+    private static final Pattern AVISUAL_NAME_PATTERN = Pattern.compile("^AVISUAL-([^/]+)$");
+
+    private final Runnable scannerRunnable;
+
+    private ScheduledFuture<?> backgroundDiscoveryFuture;
+
+    private DiscoveryServiceCallback discoveryServiceCallback;
+
+    public AirVisualNodeDiscoveryService() {
+        super(Collections.singleton(AirVisualNodeBindingConstants.THING_TYPE_AVNODE), 600, true);
+        scannerRunnable = createScannerRunnable();
+    }
+
+    @Override
+    public void setDiscoveryServiceCallback(DiscoveryServiceCallback discoveryServiceCallback) {
+        this.discoveryServiceCallback = discoveryServiceCallback;
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Starting scan");
+        scheduler.execute(scannerRunnable);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery");
+        cancelBackgroundDiscoveryFuture();
+        backgroundDiscoveryFuture = scheduler.scheduleWithFixedDelay(scannerRunnable, 0, 300, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stopping background discovery");
+        cancelBackgroundDiscoveryFuture();
+        super.stopBackgroundDiscovery();
+    }
+
+    private void cancelBackgroundDiscoveryFuture() {
+        if (backgroundDiscoveryFuture != null && !backgroundDiscoveryFuture.isDone()) {
+            backgroundDiscoveryFuture.cancel(true);
+            backgroundDiscoveryFuture = null;
+        }
+    }
+
+    private Runnable createScannerRunnable() {
+        return () -> {
+            // Get all workgroup members
+            SmbFile[] workgroupMembers;
+            try {
+                String workgroupUrl = "smb://" + AVISUAL_WORKGROUP_NAME +"/";
+                workgroupMembers = new SmbFile(workgroupUrl).listFiles();
+            } catch (Exception e) {
+                // Can't get workgroup member list
```

Comment:
Are you sure it's not needed to log here? These exceptions will go undetected and might hide bugs.